integration-tests/testsuite/src/main/java/io/apicurio/tests/utils/RateLimitingProxy.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright 2021 Red Hat
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.apicurio.tests.utils;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.vertx.core.Vertx;
+import io.vertx.core.http.HttpClient;
+import io.vertx.core.http.HttpClientOptions;
+import io.vertx.core.http.HttpClientRequest;
+import io.vertx.core.http.HttpServer;
+import io.vertx.core.http.HttpServerOptions;
+import io.vertx.core.http.HttpServerRequest;
+
+/**
+ * @author Fabian Martinez
+ */
+public class RateLimitingProxy {
+
+    protected final Logger logger = LoggerFactory.getLogger(this.getClass().getName());
+
+    private Vertx vertx;
+    private int port = 30001;
+
+    private HttpServer server;
+
+    private HttpClient client;
+    private String destinationHost;
+    private int destinationPort;
+
+    private int buckets;
+
+    public RateLimitingProxy(int failAfterRequests, String destinationHost, int destinationPort) {
+
+        // this will rate limit just based on total requests
+        // that means that if buckets=3 the proxy will successfully redirect the first 3 requests and every request after that will be rejected with 429 status
+        this.buckets = failAfterRequests;
+
+        vertx = Vertx.vertx();
+        client = vertx.createHttpClient(new HttpClientOptions());
+        if (destinationHost.endsWith("127.0.0.1.nip.io")) {
+            logger.info("Changing proxy destination host to localhost");
+            this.destinationHost = "localhost";
+        } else {
+            this.destinationHost = destinationHost;
+        }
+        this.destinationPort = destinationPort;
+    }
+
+    public String getServerUrl() {
+        return "http://localhost:" + port;
+    }
+
+    public void start() {
+
+        server = vertx.createHttpServer(new HttpServerOptions().setPort(port)).requestHandler(this::proxyRequest)
+            .listen(server -> {
+                if (server.succeeded()) {
+                    logger.info("Proxy server started on port {}", port);
+                    logger.info("Proxying server {}:{}", destinationHost, destinationPort);
+                } else {
+                    logger.error("Error starting server", server.cause());
+                }
+            });
+
+    }
+
+    public void stop() {
+        if (server != null) {
+            server.close();
+        }
+    }
+
+    private synchronized boolean allowed() {
```

Comment:
This could be implemented using `AtomicInteger`

Reply:
is ok if I merge this? we can improve that later

Reply:
sure, just a small suggestion